text/1027-ember-data-schema-service.md

```diff
@@ -0,0 +1,890 @@
+     * that this resource implements.
+     */
+    as?: string;
+
+    /**
+     * Whether this field is a polymorphic relationship,
+     * meaning that it can point to multiple types of
+     * resources so long as they implement the trait
+     * or abstract type specified in `type`.
+     */
+    polymorphic?: boolean;
+  };
+};
+
+/**
+ * > [!CAUTION]
+ * > This Field is LEGACY
+ */
+export type LegacyHasManyField = {
+  kind: 'hasMany';
+  name: string;
+  type: string;
+
+  /**
+   * Options for hasMany are mandatory.
+   */
+  options: {
+    /**
+     * Whether the relationship is async
+     *
+     * If true, it is expected that the cache
+     * data for this field will contain links
+     * or pointers that can be used to fetch
+     * the related resources when needed.
+     *
+     * When false, it is expected that all related
+     * resources are loaded together with this resource,
+     * and that the cache data for this field will
+     * contain the full list of pointers.
+     *
+     * hasMany relationships do not support pagination.
+     */
+    async: boolean;
+
+    /**
+     * The name of the inverse field on the
+     * related resource that points back to
+     * this field on this resource to form a
+     * bidirectional relationship.
+     *
+     * If null, the relationship is unidirectional.
+     */
+    inverse: string | null;
+
+    /**
+     * If this field is satisfying a polymorphic
+     * relationship on another resource, then this
+     * should be set to the trait or abstract type
+     * that this resource implements.
+     */
+    as?: string;
+
+    /**
+     * Whether this field is a polymorphic relationship,
+     * meaning that it can point to multiple types of
+     * resources so long as they implement the trait
+     * or abstract type specified in `type`.
+     */
+    polymorphic?: boolean;
+
+    /**
+     * When omitted, the cache data for this field will
+     * clear local state of all changes except for the
+     * addition of records still in the "new" state any
+     * time the remote data for this field is updated.
+     *
+     * When set to `false`, the cache data for this field
+     * will instead intelligently commit any changes from
+     * local state that are present in the remote data,
+     * leaving any remaining changes in local state still.
+     */
+    resetOnRemoteUpdate?: false;
+  };
+};
+```
+
+### Transformations
+
+Some fields have the ability to specify a transformation to serialize/deserialize the
+data for the given field when the UI accesses the value from cache or updates the value.
+
+Transformations are stateless objects that provide this ability to *hydrate* or *transform* a
+serialized cache value into a richer form for use in your App. For instance, as an Enum or Luxon
+Date.
+
+Transformations ensure that the value in the cache is always in a raw serialized form even when
+your App wants to conceptualize the state as something richer.
```

Comment:
My impression based on earlier RFCs (or discussions with you @runspired, can't remember which) was that transformations were something we were steering away from because of their fraught nature.

Looking at this though, it looks like what we're saying is the previous style of transformations was "transform results prior to storing in cache" and we're instead moving to a "transform fields on retrieval from cache for use in JS / UI". Is that an accurate statement?

Reply:
It is yes. Moving transforms to the new locations does a few things:

- they become a record concern instead of a serializer concern, in the process providing an actual guarantee of running where before they more-often-than-not would go unused
- they now prevent the cache from being polluted with non-serializable state
- they handle the create/update case where before create/update would lead to invalid cache state (strings instead of numbers etc)
- it fixes defaultValue handling
- it removes reliance on the resolver
- it allows them to function as runtime types/validations